intents/aspects.py

```diff
@@ -0,0 +1,156 @@
+
+    Returns
+        The updated table after aplying the condition, in pandas.dataframe type
+    """
+    if date_range is None:
+        return table
+    num_rows = table.shape[0]
+    for row in range(num_rows):
+        row_date = datetime.datetime.strptime(table.loc[row, date_column_name], date_format)
+        start_date = datetime.datetime.strptime(date_range[0], date_format)
+        end_date = datetime.datetime.strptime(date_range[1], date_format)
+        if row_date < start_date or end_date < row_date:
+            table = table.drop([row])
+    table = table.reset_index()
+    table = table.drop(['index'], axis=1)
+    return table
+
+def slice_table(table, slices):
+    """This function removes the rows from the table that do not satisfy the slicing condition.
+
+    Args:
+        table: Type-pandas.dataframe
+            It has the contents of the csv file
+        slices: Type-dictionary (will be changed)
+            contains the key as column name and value as instance we want to slice
+
+    Returns:
+        The updated table after aplying the condition, in pandas.dataframe type
+    """
+    if slices is None:
+    	return table
+    num_rows = table.shape[0]
+    for row in range(num_rows):
+        slice_match = True
+        for key in slices:
+            if table.loc[row, key] != slices[key]:
+                slice_match = False
+        if slice_match is not True:
+            table = table.drop([row])
+
+    table = table.reset_index()    # some indices get deleted after slicing
+    table = table.drop(['index'], axis=1)    # droping the new columnn named 'index' created after reset_index call
+    return table
+
+def corp_other_dimensions(table, dimensions, metric):
+    """This function removes the columns that are not in dimensions and are not equal to metric
+
+    Args:
+        table: Type-pandas.dataframe
+        dimensions: Type-list of str
+        metric: Type-string
+
+    Returns:
+        Returns the table as a dataframe obj after removing the rest of the columns.
+    """
+
+    current_columns = table.columns
+    for column in current_columns:
+        if (column != metric) and (column not in dimensions) :
+            table = table.drop([column], axis=1)
+
+    return table
+
+def count_distinct(values):
+    """ This function returns the number of distinct entries in values.
+    The values are inserted into a set and it's size is returned.
+
+    Args:
+        values: Type - anything that can be iterated over
+
+    Returns:
+        Count of distinct values.
+    """
+    distinct_set = set()
+    for value in values:
+        distinct_set.add(value)
+    return len(distinct_set)
+
+def group_by(table, dimensions, summary_operator):
+    """Groups the column by the columns in dimensions
+    Basically makes a map in which keys contain dimensions and values are evaluated after applying the summary operator.
```

Comment:
How does grouping handles empty values in the dimension? Does it consider them as a group or ignores such rows?

Reply:
It actually drops every row which contain NaN in the list of rows that we group by.
But it can be easily customized.